heartbeat/beater/heartbeat.go

```diff
@@ -56,36 +64,80 @@ func New(b *beat.Beat, cfg *common.Config) (beat.Beater, error) {
 		return nil, err
 	}
 
-	sched := scheduler.NewWithLocation(limit, location)
-	manager, err := newMonitorManager(b.Publisher, sched, monitors.Registry, config.Monitors)
+	scheduler := scheduler.NewWithLocation(limit, location)
 	if err != nil {
 		return nil, err
 	}
 
 	bt := &Heartbeat{
-		done:      make(chan struct{}),
-		scheduler: sched,
-		manager:   manager,
+		done:           make(chan struct{}),
+		config:         parsedConfig,
+		scheduler:      scheduler,
+		monitorFactory: monitors.NewFactory(scheduler),
 	}
 	return bt, nil
 }
 
+// Run executes the beat.
 func (bt *Heartbeat) Run(b *beat.Beat) error {
 	logp.Info("heartbeat is running! Hit CTRL-C to stop it.")
 
+	err := bt.RunStaticMonitors(b)
+	if err != nil {
+		return err
+	}
+
+	onDone, err := bt.RunDynamicMonitors(b)
+	if err != nil {
+		return err
+	}
+
 	if err := bt.scheduler.Start(); err != nil {
 		return err
 	}
 	defer bt.scheduler.Stop()
 
 	<-bt.done
-
-	bt.manager.Stop()
+	if onDone != nil {
+		onDone()
+	}
 
 	logp.Info("Shutting down.")
 	return nil
 }
 
+// RunStaticMonitors runs the `heartbeat.monitors` portion of the yaml config if present.
+func (bt *Heartbeat) RunStaticMonitors(b *beat.Beat) error {
+	for _, cfg := range bt.config.Monitors {
+		created, err := bt.monitorFactory.Create(b.Publisher, cfg, nil)
+		if err != nil {
+			return errors.Wrap(err, "could not create monitor")
+		}
+		created.Start()
+	}
+	return nil
+}
+
+// RunDynamicMonitors runs the `heartbeat.config.monitors` portion of the yaml config if present.
+func (bt *Heartbeat) RunDynamicMonitors(b *beat.Beat) (onDone func(), err error) {
```

Comment:
consider to return the monitorReloader or pass the monitorReloader.

E.g. in beater Run:

```
if bt.config.ConfigMonitors.Enabled() {
  reloader := cfgfile.NewReloader(...)
  err := startReloader(reloader, ...)
  if err != nil { ... }

  defer reloader.Stop() // defer has function scope
}
```

with 

```
func startReloader(reloader, factory) error {
  if err := reloader.Check(factory); err != nil {
     return err
  }
  go reloader.Run(factory)
  return nil
}
```

Reply:
Ah, thanks for the heads up on that, makes sense, will change it